tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Hinge.java

```diff
@@ -124,16 +124,13 @@ public Hinge(Ops tf, String name, Reduction reduction) {
   public <T extends TNumber, U extends TNumber> Operand<T> call(
       Operand<U> labels, Operand<T> predictions, Operand<T> sampleWeights) {
```

Comment:
`U extends TFloating`?

Reply:
I did not saw any checks enforcing that the tensors to be of a floating type for `Hinge`, should it?

Reply:
Well it says predictions should be between 0 and 1. I guess it could supply the integers rather than probabilities, but it's kinda against the docs.

Reply:
Labels can be -1, or 1, if the are 0,1, they are converted to -1, 1, but python example show this as float.
Python is silent on the Predictions, but all the values are between 0 and 1, so should probably be TFLoating.

Reply:
The same types should probably apply to SquaredHinge(labels = -1,1 or 0,1) and CategoricalHinge (labels = 0,1)

Reply:
Just to make sure I understood @JimClarke5 , so labels could be any numeric values while predictions and sampleWeights must be restricted to floating-points only in all three losses you've mentioned?

Reply:
For these classes, Python does not dictate that they be integers, just that they are value restricted to -1, 0, 1. 
I think if we restrict it to TInteger, it falls more in line with the restriction on the values.

Reply:
Ok so then labels (`U`) is bound to `TIntegral` and predictions/weights (`T`) is bound to `TFloating`, I will make this change and we'll adjust later if needed.

Reply:
But in order to change the bound of these generics, I need to move the parameter at the `Loss` class level instead of the `call` method level, otherwise the Java compiler does not allow me to override them. e.g.
```java
public abstract class Loss<T, U> {
    public abstract Operand<T> call(Operand<U> labels, Operand<T> predictions, Operand<T> sampleWeights);
}

public final class Hinge extends Loss<TFloating, TIntegral> {
    @Override
    public abstract Operand<T> call(Operand<U> labels, Operand<T> predictions, Operand<T> sampleWeights) {
    // ...
    }
}
```
Are we ok with this change or should we leave `Loss` parameter-less? ... or should we simply leave it as is (i.e. only bound to `TNumber`) and review this in a separate PR?

Reply:
I don't understand what you need to do? With ` Hinge extends Loss<TFloating, TIntegral>`,
Why doesn't `call(Operand<TInt32> labels, Operand<TFloat32> predictions, Operand<TFloat32> sampleWeigthts)` work?

Reply:
That's the thing, it would work that way. But right now the `Loss` class is unparameterized and the parameters are carried at the [method level](https://github.com/tensorflow/java/blob/3f3f384517a5ba633c3b1971651cdcdec3eebfa7/tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Loss.java#L68). I'm not able to override the generic bounds at the method level, only at the class level but that would require me to do a bit of refactoring like I've presented above.

Reply:
Ok, now I see. I would change `Loss` to to be parameterized. That will require touch all the `Loss` subclasses though, but I think that makes more sense.

Reply:
BTW: Metric is parameterized, so I guess I am learning :-).

Reply:
In the interests of getting this PR integrated, lets leave firming up these types for another PR.

Reply:
I agree, so I'm pushing only the other changes that we've discussed.